contracts/farm/Farm.sol

```diff
@@ -0,0 +1,197 @@
+    using SafeToken for address;
+
+    struct StakeInfo {
+        uint256 amount;
+        uint256 rewardDebt;
+    }
+
+    struct PoolInfo {
+        uint256 allocPoint;
+        uint256 lastRewardBlock;
+        uint256 accRewardPerShare;
+    }
+
+    address public rewardToken;
+
+    uint256 public rewardPerBlock;
+    uint256 public totalAllocPoint;
+
+    mapping(address => PoolInfo) public poolInfos;
+    mapping(address => mapping(address => StakeInfo)) public stakes;
+    mapping(address => uint256) private lpTokenBalances;
+
+    event LogPoolAdded(address indexed lpToken, uint256 allocPoint);
+    event LogPoolAllocationUpdate(address indexed lpToken, uint256 allocPoint);
+    event LogPoolRewardUpdate(
+        address indexed lpToken,
+        uint256 accRewardPerShare,
+        uint256 lastRewardBlock
+    );
+    event LogDeposit(address indexed user, address indexed lpToken, uint256 amount);
+    event LogWithdraw(address indexed user, address indexed lpToken, uint256 amount);
+    event LogRewardPaid(address indexed user, address indexed lpToken, uint256 reward);
+
+    function initialize(
+        address _rewardToken,
+        uint256 _rewardPerBlock
+    ) external initializer {
+        OwnableUpgradeable.__Ownable_init();
+        PausableUpgradeable.__Pausable_init();
+
+        rewardToken = _rewardToken;
+        rewardPerBlock = _rewardPerBlock;
+        totalAllocPoint = 0;
+    }
+
+    function addPool(
+        address _lpToken,
+        uint256 _allocPoint
+    ) external override onlyOwner whenNotPaused {
+        poolInfos[_lpToken] = PoolInfo({
+            allocPoint: _allocPoint,
+            lastRewardBlock: block.number,
+            accRewardPerShare: 0
+        });
+
+        totalAllocPoint += _allocPoint;
+
+        emit LogPoolAdded(_lpToken, _allocPoint);
+    }
+
+    function updateAllocation(
+        address _lpToken,
+        uint256 _allocPoint
+    ) external override onlyOwner whenNotPaused {
+        PoolInfo storage pool = poolInfos[_lpToken];
+
+        totalAllocPoint = totalAllocPoint + _allocPoint - pool.allocPoint;
+        pool.allocPoint = _allocPoint;
+
+        emit LogPoolAllocationUpdate(_lpToken, _allocPoint);
+    }
+
+    function deposit(address _lpToken, uint256 _amount) external override whenNotPaused {
+        require(_lpToken != address(0), "Zero address");
+        require(_amount > 0, "Zero amount");
+
+        updatePoolReward(_lpToken);
+
+        PoolInfo storage pool = poolInfos[_lpToken];
+        StakeInfo storage stake = stakes[_lpToken][msg.sender];
+
+        if (stake.amount > 0) {
+            uint256 pending = (stake.amount * pool.accRewardPerShare) /
+                1e12 -
+                stake.rewardDebt;
+            if (pending > 0) {
+                address(rewardToken).safeTransfer(msg.sender, pending);
+                emit LogRewardPaid(msg.sender, _lpToken, pending);
+            }
```

Comment:
not sure if reward payment on deposit is understandable behaviour for the user. we aren't making restaking of the rewards, but lets make them withdrawn in a separate function